pkg/volume/downwardapi/downwardapi.go

```diff
@@ -0,0 +1,370 @@
+		// No data changed: nothing to write
+		return nil
+	}
+
+	if err := b.writeData(data); err != nil {
+		glog.Errorf("Unable to dump files for downwardAPI volume %v for pod %v/%v: %s", b.volName, b.pod.Namespace, b.pod.Name, err.Error())
+		return err
+	}
+
+	glog.V(3).Infof("Data dumped for downwardAPI volume %v for pod %v/%v", b.volName, b.pod.Namespace, b.pod.Name)
+	return nil
+}
+
+// IsReadOnly func to fullfill volume.Builder interface
+func (d *downwardAPIVolume) IsReadOnly() bool {
+	return true
+}
+
+// collectData collects requested downwardAPI in data map.
+// Map's key is the requested name of file to dump
+// Map's value is the (sorted) content of the field to be dumped in the file.
+func (d *downwardAPIVolume) collectData() (map[string]string, error) {
+	errlist := []error{}
+	data := make(map[string]string)
+	for fieldReference, fileName := range d.fieldReferenceFileNames {
+		if values, err := fieldpath.ExtractFieldPathAsString(d.pod, fieldReference); err != nil {
+			glog.Errorf("Unable to extract field %s: %s", fieldReference, err.Error())
+			errlist = append(errlist, err)
+		} else {
+			data[fileName] = sortLines(values)
+		}
+	}
+	return data, utilErrors.NewAggregate(errlist)
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (d *downwardAPIVolume) isDataChanged(data map[string]string) bool {
+	for fileName, values := range data {
+		if isFileToGenerate(path.Join(d.GetPath(), fileName), values) {
+			return true
+		}
+	}
+	return false
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const (
+	downwardAPIDir    = "..downwardapi"
+	downwardAPITmpDir = "..downwardapi_tmp"
+)
+
+// writeData writes requested downwardAPI in specified files.
+//
+// The file visible in this volume are symlinks to files in the '.downwardapi'
+// directory. Actual files are stored in an hidden timestamped directory which is
+// symlinked to by '.downwardapi'. The timestamped directory and '.downwardapi' symlink
+// are created in the plugin root dir.  This scheme allows the files to be
+// atomically updated by changing the target of the '.downwardapi' symlink.  When new
+// data is available:
+//
+// 1.  A new timestamped dir is created by writeDataInTimestampDir and requested data
+//     is written inside new timestamped directory
+// 2.  Symlinks and directory for new files are created (if needed).
+//     For example for files:
+//       <volume-dir>/user_space/labels
+//       <volume-dir>/k8s_space/annotations
+//       <volume-dir>/podName
+//     This structure is created:
+//       <volume-dir>/podName               -> ..downwardapi/podName
+//       <volume-dir>/user_space/labels     -> ../..downwardapi/user_space/labels
```

Comment:
Those are the relative links for files in sub directories... maybe we can add a note about that @sdminonne